pkg/distributor/distributor.go

```diff
@@ -605,9 +609,87 @@ func (d *Distributor) validateSeries(nowt time.Time, ts mimirpb.PreallocTimeseri
 	return nil
 }
 
-// PrePushForwardingMiddleware is used as push.Func middleware in front of PushWithCleanup method.
+func (d *Distributor) pushWithMiddlewares() push.Func {
+	var middlewares []func(push.Func) push.Func
+
+	// The middlewares will be applied in the order of the slice "middlewares",
+	// requests will traverse them in the reverse order.
+	middlewares = append(middlewares, d.prePushForwardingMiddleware)
+	middlewares = append(middlewares, d.prePushHaDedupeMiddleware)
+
+	push := d.PushWithCleanup
+	for _, middleware := range middlewares {
+		push = middleware(push)
+	}
+
+	return push
+}
+
+func (d *Distributor) prePushHaDedupeMiddleware(next push.Func) push.Func {
+	return func(ctx context.Context, req *mimirpb.WriteRequest, cleanup func()) (*mimirpb.WriteResponse, error) {
+		userID, err := tenant.TenantID(ctx)
+		if err != nil {
+			cleanup()
```

Comment:
I'd suggest to use the pattern from `PushWithCleanup`:

```
	cleanupInDefer := true
	defer func() {
		if cleanupInDefer {
			cleanup()
		}
	}()
```

and only set `cleanupInDefer = false` before calling `next`. It's easier (fewer places) than to check if all returns are covered.

Reply:
yeah, fair enough, I'll update that

Reply:
how is this: d0289e4a0

Reply:
Thanks!